docs/source/extensions_tutorial/3_spec_api.rst

```diff
@@ -0,0 +1,179 @@
+
+.. tip::
+    Although you have the option not to, there are several advantages to defining new groups and neurodata types.
+    Neurodata types can be reused in multiple places in the schema, and can be linked to, while groups that are not
+    neurodata types cannot. You can also have multiple neurodata type groups of the same type in the same group,
+    whereas groups that are not neurodata types are limited to 0 or 1. Most of the time, we would recommend making a
+    group a neurodata type. It is also generally better to extend your neurodata type from an existing type. Look
+    through the `NWB schema <https://nwb-schema.readthedocs.io/en/latest/>`_ to see if a core neurodata type would
+    work as a base for your new type. If no existing type works, consider extending
+    :py:class:`~pynwb.base.NWBDataInterface``, which allows you to add the object to a processing module.
+
+.. tip::
+     New neurodata types should always be declared at the top level of the schema rather than nesting type
+     definitions. I.e., when creating a new neurodata type it should be placed at the top level of your schema
+     and then included at the appropriate location via ``neurodata_type_inc``. This approach greatly simplifies
+     management of types.
+
+For more information about the options available when specifying a Group, see the
+`API docs for NWBGroupSpec <https://pynwb.readthedocs.io/en/stable/pynwb.spec.html?highlight=NWBGroupSpec#pynwb.spec.NWBGroupSpec>`_.
+
+Dataset Specifications
+^^^^^^^^^^^^^^^^^^^^^^
+
+All larger blocks of numeric or text data should be stored in Datasets. Specifying datasets is done with
+:py:class:`~pynwb.spec.NWBDatasetSpec`.
+
+.. code-block:: python
+
+    from pynwb.spec import NWBDatasetSpec
+
+    spec = NWBDatasetSpec(
+        doc='A custom NWB type',
+        name='qux',
+        shape=(None, None),
+        attributes=[...]
+    )
+
+``neurodata_type_def``, ``neurodata_type_inc``, ``doc``, ``name``, ``default_name``, ``linkable``, ``quantity``, and
+``attributes`` all work the same as they do in :py:class:`~pynwb.spec.NWBGroupSpec`, described in the previous section.
+
+``dtype`` defines the type of the data, which can be a basic type or compound or reference type. 
+See a list of options`here <https://schema-language.readthedocs.io/en/latest/description.html#dtype>`_. 
+Basic types can be defined as string objects and more complex types via :py:class:`~pynwb.spec.NWBDtypeSpec` and `RefSpec <https://hdmf.readthedocs.io/en/latest/hdmf.spec.spec.html#hdmf.spec.spec.RefSpec>`_.
+
+
+``shape`` is a specification defining the allowable shapes for the dataset. See the shape specification
+`here <https://schema-language.readthedocs.io/en/latest/specification_language_description.html#shape>`_. ``None`` is
+mapped to ``null``. Is no shape is provided, it is assumed that the dataset is only a single element.
+
+``dims`` provides labels for each dimension of ``shape``.
+
+``default_value`` is also available.
+
```

Comment:
This is rarely used and has caveats (does not work if the dataset is not a scalar), so I recommend omitting this.
```suggestion
```